### PR TITLE
Dispose root context before unloading process.

### PR DESCRIPTION
**Pull request by michaelcfanning**
This change ensures that every 'root' context that is created at analysis time is disposed in all code paths. This hasn't been a problem to-date as no tool has required an active dispose before process unload. BinSkim now collects telemetry that is finalized in a telemetry logger Dispose method, however, and so it must be definitively dispose on process exit to ensure all data is flushed and written.

@marmegh @shaopeng-gh @eddynaka @EasyRhinoMSFT @yongyan-gh

### CODE REVIEW

File: src/Sarif/Core/SarifLog.cs [hunk 1]
```diff
@@ -87,7 +87,7 @@ public static SarifLog Load(Stream source, bool deferred = false)
 
             if (!fileSystem.FileExists(filePath))
             {
-                throw new ArgumentException(nameof(filePath));
+                throw new ArgumentException($"File path does not exist: '{filePath}'", nameof(filePath));
```
Comment (michaelcfanning) [inline src/Sarif/Core/SarifLog.cs:90]: >ArgumentException [](http://example.com/codeflow?start=26&length=17)

This is a very comment API use issue, this exception type does not accept a method overload that only references the parameter name (unlike, say, NullReferenceException).

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs [hunk 2]
```diff
@@ -66,71 +66,78 @@ public string DefaultConfigurationPath
...
                 }
-            }
 
-            if (options.RichReturnCode)
+                return succeeded ? SUCCESS : FAILURE;
+            }
+            finally
             {
-                return (int)RuntimeErrors;
+                _rootContext?.Dispose();
```
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:139]: >_rootContext?.Dispose [](http://example.com/codeflow?start=16&length=21)

Here's the key change. In all code paths on calling Run, we dispose the root context object if non-null.

File: src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs [hunk 3]
```diff
@@ -1237,10 +1250,10 @@ private void AnalyzeScenarios(int[] scenarios)
 
                 testCase.FileSystem = null;
                 testCase.Files = multiThreadTargets;
-                Run runMultiThread = RunAnalyzeCommand(options, testCase, multithreaded: true);
+                Run runMultithreaded = RunAnalyzeCommand(options, testCase, multithreaded: true);
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs:1253]: >runMultithreaded [](http://example.com/codeflow?start=20&length=16)

Notice that I fix incidental code issues that I notice while I'm working. My goal is that *every* code change leaves the code base in a slightly better condition, even when I am extending/adding functionality to it.

File: src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs [hunk 4]
```diff
@@ -40,6 +40,19 @@ public AnalyzeCommandBaseTests(ITestOutputHelper output)
             this.Output = output;
         }
 
+        [Fact]
+        public void AnalyzeCommandBase_RootContextIsDisposed()
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs:44]: >AnalyzeCommandBase_RootContextIsDisposed [](http://example.com/codeflow?start=20&length=40)

Observe my test approach. First I add a focused test for both single and multithreaded analysis. This test, however, does not come close to testing all possible exit paths (this code returns in various places to handle exceptional paths, for example). So I will go inject this validation directly into our test commands as well.

This is a general strategy of mine when adding tests to legacy code: author directed tests, but look for opportunities to generalize coverage in the test infrastructure, via dependency injection, etc.

File: src/Test.UnitTests.Sarif.Driver/TestMultithreadedAnalyzeCommand.cs [hunk 5]
```diff
@@ -59,7 +61,9 @@ protected override void ValidateOptions(TestAnalyzeOptions options, TestAnalysis
 
         public int Run(AnalyzeOptionsBase options)
         {
-            return base.Run((TestAnalyzeOptions)options);
+            int result = base.Run((TestAnalyzeOptions)options);
+            this._rootContext?.Disposed.Should().BeTrue();
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif.Driver/TestMultithreadedAnalyzeCommand.cs:65]: >_rootContext [](http://example.com/codeflow?start=17&length=12)

Now every test that travels this code path will enforce this invariant.

File: src/Test.UnitTests.Sarif.Driver/TestAnalyzeCommand.cs [hunk 6]
```diff
@@ -70,5 +72,12 @@ protected override void ProcessBaseline(IAnalysisContext context, TestAnalyzeOpt
 
             base.ProcessBaseline(context, options, fileSystem);
         }
+
+        public override int Run(TestAnalyzeOptions options)
+        {
+            int result = base.Run(options);
+            this._rootContext?.Disposed.Should().BeTrue();
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif.Driver/TestAnalyzeCommand.cs:79]: >_rootContext [](http://example.com/codeflow?start=17&length=12)

Now every test that travels this code path will enforce this invariant.

**Review by eddynaka — APPROVED** [review]
:shipit:

**Pull request merged**